core_lib/src/managers/backupmanager.cpp

```diff
@@ -0,0 +1,711 @@
+        mBackupAtSave = latestBackupElement();
+    } else {
+        mLegacyBackupAtSave = mLegacyBackupList[mLegacyBackupIndex];
+    }
+    return Status::OK;
+}
+
+void BackupManager::backup(BackupType type)
+{
+    if (!newUndoRedoSystemEnabled()) {
+        return;
+    }
+
+    Layer* currentLayer = editor()->layers()->currentLayer();
+    switch (type)
+    {
+        case BackupType::STROKE:
+        {
+            if (currentLayer->type() == Layer::BITMAP) {
+                bitmap(tr("Bitmap stroke"));
+            } else if (currentLayer->type() == Layer::VECTOR) {
+                vector(tr("Vector stroke"));
+            } else {
+                Q_ASSERT_X(false, "BackupManager", "A stroke can only be applied to either the Bitmap or Vector layer");
+            }
+            break;
+        }
+        case BackupType::POLYLINE:
+        {
+            if (currentLayer->type() == Layer::BITMAP) {
+                bitmap(tr("Bitmap polyline"));
+            } else if (currentLayer->type() == Layer::VECTOR) {
+                vector(tr("Vector polyline"));
+            } else {
+                Q_ASSERT_X(false, "BackupManager", "A polyline can only be applied to either the Bitmap or Vector layer");
+            }
+            break;
+        }
+        case BackupType::SELECTION:
+        {
+            if (currentLayer->type() == Layer::BITMAP) {
+                selection(tr("Bitmap selection"));
+            } else if (currentLayer->type() == Layer::VECTOR) {
+                selection(tr("Vector selection"));
+            } else {
+                Q_ASSERT_X(false, "BackupManager", "A polyline can only be applied to either the Bitmap or Vector layer");
+            }
+            break;
+        }
+        default:
+            Q_ASSERT_X(false, "BackupManager", "Tried to make a backup for a case which hasn't been handled yet");
+    }
+}
+
+const BackupElement* BackupManager::latestBackupElement() const
+{
+    return static_cast<const BackupElement*>(mUndoStack->command(mUndoStack->index() - 1));
+}
+
+bool BackupManager::newUndoRedoSystemEnabled() const
+{
+    return mNewBackupSystemEnabled;
+}
+
+bool BackupManager::hasUnsavedChanges() const
+{
+    if (newUndoRedoSystemEnabled()) {
+        return mBackupAtSave != latestBackupElement();
+    } else {
+        if (mLegacyBackupIndex >= 0) {
+            return mLegacyBackupAtSave != mLegacyBackupList[mLegacyBackupIndex];
+        }
+        return false;
+    }
+}
+
+void BackupManager::pushCommand(QUndoCommand* command)
+{
+    mUndoStack->push(command);
+    emit didUpdateUndoStack();
+}
+
+void BackupManager::bitmap(const QString& description)
+{
+    if (mBitmap == nullptr) { return; }
+    BitmapElement* element = new BitmapElement(mBitmap,
+                                               mLayerId,
+                                               editor(),
+                                               description);
+
+    new TransformElement(mKeyframe,
```

Comment:
That's because TransformElement has to perform keyframe related modifications, just like BitmapElement, they're not as such able to use each other. You can technically get access to the child element, for example the TransformElement in BitmapElement but since TransformElement can also be used in a context where it does not have a BitmapElement, it needs the Keyframe anyway.